packages/blocks/src/api/test/parser.js

```diff
@@ -529,10 +532,53 @@ describe( 'block parser', () => {
 				],
 			} );
 
-			const migratedBlock = getMigratedBlock( block );
+			const migratedBlock = getMigratedBlock( block, parsedAttributes );
 
 			expect( migratedBlock.attributes ).toEqual( { fruit: 'Bananas!' } );
 		} );
+
+		it( 'allows a default attribute to be deprecated', () => {
```

Comment:
Being a bit confused about what actually changed as part of the implementation aside from moving attributes into an argument rather than having it be destructured, I copied this test and this test only into master. It passed. I can't imagine that's expected? I'd expect the test case would be one which only passes in the branch, and fails on master.

Reply:
Thanks for the catch - looks like I made a mistake in the test with attribute type, which was causing a false positive (the types didn't match so that attribute was being reset). Fixed here:
https://github.com/WordPress/gutenberg/pull/12757/commits/ec0e490375ed3de842a05e28e421e37b5f6ff1f8

> Being a bit confused about what actually changed as part of the implementation aside from moving attributes into an argument rather than having it be destructured.

It's really hard to explain!

The difference is that before this change, the attributes may have had defaults applied from the main block type:
https://github.com/WordPress/gutenberg/blob/master/packages/blocks/src/api/parser.js#L455

If the main block type fails validation, the deprecations are attempted, but the default value from main block type is passed in when testing the deprecations, which doesn't give the deprecations a chance to apply their own defaults.

My change instead passes the original attributes from before those defaults from the main block type have been applied into `getMigratedBlock`, which gives the deprecations a chance to use their defaults.